classes/local/lib.php

```diff
@@ -139,8 +138,27 @@ public static function cm_info_dynamic(\cm_info $cminfo) {
      */
     public static function cm_info_view(\cm_info $cminfo) {
         $cms = new cms($cminfo->instance);
-        $renderer = new renderer($cms);
-        $cminfo->set_content($renderer->get_html());
+
+        $cachekey = 'super_hash_' . $cminfo->instance;
+        $hashcache = \cache::make('mod_cms', 'datasource_keys');
+        $storedhash = $hashcache->get($cachekey);
+        $contentcache = \cache::make('mod_cms', 'content_cache');
+
+        $currenthash = '';
+        foreach (dsbase::get_datasources($cms) as $ds) {
+            $currenthash .= $ds->get_cache_hash();
+        }
+
+        if ($storedhash !== false && $currenthash === $storedhash) {
+            $content = $contentcache->get($cminfo->instance);
+        } else {
+            $renderer = new renderer($cms);
+            $content = $renderer->get_html();
+            $contentcache->set($cminfo->instance, $content);
+            $hashcache->set($cachekey, $currenthash);
```

Comment:
Would this be cleaner if it was moved up after line 144, these two hashes should be independent shouldn't they?

I'm a little confused about what is going on with this, which might just be a code commenting thing. It's not clear to me why we need 2 caches.

I would be expecting $contentcache to be getting and setting a value based on the $currenthash and not on the $cminfo->instance, ie if you happened to create two instances which had the same data then they should have the same hash keys and use the same cache value. If you change the data then you create a new hash key which uses a different cache slot which is how you get the localized caches working.

get_cache_hash in my mind should return something which changes when the data changes, so it could be a timestamp or it could be a sha1 hash of the instance data, it itself I don't think should be calling MUC.